src/extensions/default/Inline3DParametersEditor/InlineParameterEditor.js

```diff
@@ -0,0 +1,213 @@
+            }
+
+            // Don't push the change back into the host editor if it came from the host editor.
+            if (!this._isHostChange) {
+                var endPos = {
+                        line: range.start.line,
+                        ch: range.start.ch + parameterString.length
+                    };
+                this._isOwnChange = true;
+                this.hostEditor.document.batchOperation(function () {
+                    self.hostEditor.setSelection(range.start, range.end); // workaround for #2805
+                    self.hostEditor.document.replaceRange(parameterString, range.start, range.end, self._origin);
+                    self.hostEditor.setSelection(range.start, endPos);
+                    if (self._marker) {
+                        self._marker.clear();
+                        self._marker = self.hostEditor._codeMirror.markText(range.start, endPos);
+                    }
+                });
+                this._isOwnChange = false;
+            }
+
+            this._parameters = parameterString;
+        }
+    };
+
+    /**
+     * @override
+     * @param {!Editor} hostEditor
+     */
+    InlineParameterEditor.prototype.load = function (hostEditor) {
+        InlineParameterEditor.prototype.parentClass.load.apply(this, arguments);
+        this.parameterEditor = new ParameterEditor(this.$htmlContent, this._handleParametersChange, this._tag, this._parameters);
+    };
+
+    /**
+     * @override
+     * Perform sizing & focus once we've been added to Editor's DOM
+     */
+    InlineParameterEditor.prototype.onAdded = function () {
+        InlineParameterEditor.prototype.parentClass.onAdded.apply(this, arguments);
+
+        var doc = this.hostEditor.document;
+        doc.addRef();
+        doc.on("change", this._handleHostDocumentChange);
+
+        this.hostEditor.setInlineWidgetHeight(this, this.parameterEditor.getRootElement().outerHeight(), true);
+
+        this.parameterEditor.focus();
+    };
+
+    /**
+     * @override
+     * Called whenever the inline widget is closed, whether automatically or explicitly
+     */
+    InlineParameterEditor.prototype.onClosed = function () {
+        InlineParameterEditor.prototype.parentClass.onClosed.apply(this, arguments);
+
+        if (this._marker) {
+            this._marker.clear();
+        }
+
+        var doc = this.hostEditor.document;
+        doc.off("change", this._handleHostDocumentChange);
+        doc.releaseRef();
+    };
+
+    /**
+     * @override
+     * Called whenever the inline widget is closed, whether automatically or explicitly
+     */
+    InlineParameterEditor.prototype.onClosed = function () {
+        InlineParameterEditor.prototype.parentClass.onClosed.apply(this, arguments);
+
+        if (this._marker) {
+            this._marker.clear();
+        }
+
+        var doc = this.hostEditor.document;
+        doc.off("change", this._handleHostDocumentChange);
+        doc.releaseRef();
```

Comment:
Lines 182-188 are duplicated here and in 166-172. Let's move this to a function you call in both places.